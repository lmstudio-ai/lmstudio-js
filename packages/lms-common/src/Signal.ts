import { type Patch } from "@lmstudio/immer-with-plugins";
import { type StripNotAvailable } from "./LazySignal.js";
import { Subscribable } from "./Subscribable.js";
import { makePromise } from "./makePromise.js";
import { makeSetterWithPatches, type Setter, type WriteTag } from "./makeSetter.js";

const equals = <TValue>(a: TValue, b: TValue) => a === b;

type Updater<TValue> = (oldValue: TValue) => readonly [TValue, Array<Patch>];
export type Subscriber<TValue> = (value: TValue) => void;

export type SignalFullSubscriber<TValue> = (
  value: TValue,
  patches: Array<Patch>,
  tags: Array<WriteTag>,
) => void;

type InternalSubscriber<TValue> =
  | {
      type: "regular";
      callback: Subscriber<TValue>;
    }
  | {
      type: "full";
      callback: SignalFullSubscriber<TValue>;
    };

/**
 * A signal is a wrapper for a value. It can be used to notify subscribers when the value changes.
 * For it to work properly, the value should be immutable.
 *
 * To create a signal, please use the `Signal.create` static method. It will return a signal
 * along with a function to update its value.
 */
export class Signal<TValue> extends Subscribable<TValue> implements SignalLike<TValue> {
  /**
   * Creates a signal.
   *
   * @param value - The initial value of the signal.
   * @param equalsPredicate - A function to compare two values. The subscribers will only be called
   * if the value changes according to the `equalsPredicate`. By default, it uses the `===`
   * operator.
   * @returns This method returns a tuple with two elements:
   * - The signal
   * - A function to update the value
   **/
  public static create<TValue>(
    value: TValue,
    equalsPredicate: (a: TValue, b: TValue) => boolean = equals,
  ): readonly [Signal<TValue>, Setter<TValue>] {
    const signal = new Signal(value, equalsPredicate);
    const update = (updater: Updater<TValue>, tags?: Array<WriteTag>) => {
      signal.update(updater, tags);
    };
    const setter = makeSetterWithPatches(update);
    return [signal, setter] as const;
  }
  public static createReadonly<TValue>(value: TValue): Signal<TValue> {
    return Signal.create(value)[0];
  }
  protected constructor(
    private value: TValue,
    private equalsPredicate: (a: TValue, b: TValue) => boolean,
  ) {
    super();
  }
  private subscribers: Set<InternalSubscriber<TValue>> = new Set();
  /**
   * Returns the current value of the signal.
   */
  public get() {
    return this.value;
  }
  public pull() {
    return this.value as StripNotAvailable<TValue>;
  }
  private queuedUpdaters: Array<[updater: Updater<TValue>, tags?: Array<WriteTag>]> = [];
  private isEmitting = false;
  private notifyFull(value: TValue, patches: Array<Patch>, tags: Array<WriteTag>) {
    for (const { type, callback } of this.subscribers) {
      if (type === "full") {
        callback(value, patches, tags);
      }
    }
  }
  private notifyAll(value: TValue, patches: Array<Patch>, tags: Array<WriteTag>) {
    for (const { type, callback } of this.subscribers) {
      if (type === "regular") {
        callback(value);
      } else {
        callback(value, patches, tags);
      }
    }
  }
  private notifyAndUpdateIfChanged(value: TValue, patches: Array<Patch>, tags: Array<WriteTag>) {
    // If the value has changed, or if there are any tags that need to be flushed, notify
    if (!this.equalsPredicate(this.value, value)) {
      this.value = value;
      // If the values have changed, notify everyone
      this.notifyAll(value, patches, tags);
    } else if (tags.length > 0) {
      // If values not changed, but there is a tag to be flushed, notify only full subscribers
      this.notifyFull(value, patches, tags);
    }
  }

  private isReplaceRoot(patch: Patch) {
    return patch.path.length === 0 && patch.op === "replace";
  }

  private update(updater: Updater<TValue>, tags?: Array<WriteTag>) {
    this.queuedUpdaters.push([updater, tags]);
    // Only one concurrent update may emit
    if (this.isEmitting) {
      return;
    }
    this.isEmitting = true;
    try {
      // Outer while is for handling new updates caused by the notify
      while (this.queuedUpdaters.length > 0) {
        let value = this.value;
        let patches: Array<Patch> = [];
        const tags: Array<WriteTag> = [];
        // Inner while is for handling multiple updates
        while (this.queuedUpdaters.length > 0) {
          const [updater, newTags] = this.queuedUpdaters.shift()!;
          const [newValue, newPatches] = updater(value);
          value = newValue;
          // Extremely rudimentary patch merging
          const rootReplacerIndex = newPatches.findIndex(this.isReplaceRoot);
          if (rootReplacerIndex !== -1) {
            patches = newPatches.slice(rootReplacerIndex);
          } else {
            patches.push(...newPatches);
          }
          if (newTags !== undefined) {
            tags.push(...newTags);
          }
        }
        this.notifyAndUpdateIfChanged(value, patches, tags);
      }
    } finally {
      this.isEmitting = false;
    }
  }

  /**
   * Subscribes to the signal. The callback will be called whenever the value changes. All callbacks
   * are called synchronously upon updating. It will NOT be immediately called with the current
   * value. (Use `get()` to get the current value.) Returns a function to unsubscribe.
   *
   * Edge cases involving manipulating the signal in the callback:
   *
   * - If the callback adds new subscribers, they will also be called within the same update.
   * - If the callback causes removal of subscribers that have not been called yet, they will no
   *   longer be called.
   * - If the callback causes an update of the value, the update will be queued. If multiple updates
   *   are queued, only the last one will be executed.
   *
   * Edge cases involving adding the same callback multiple times.
   *
   *  - Callbacks are tracked with a set. Adding the same subscriber will not cause it to be called
   *    multiple times.
   */
  public subscribe(callback: Subscriber<TValue>): () => void {
    const subscriber: InternalSubscriber<TValue> = {
      type: "regular",
      callback,
    };
    this.subscribers.add(subscriber);
    return () => {
      this.subscribers.delete(subscriber);
    };
  }

  /**
   * Subscribes to the signal with the callback and trigger the callback immediately with the
   * current value.
   */
  public subscribeAndNow(callback: Subscriber<TValue>): () => void {
    const unsubscribe = this.subscribe(callback);
    callback(this.value);
    return unsubscribe;
  }

  public subscribeFull(callback: SignalFullSubscriber<TValue>): () => void {
    const subscriber: InternalSubscriber<TValue> = {
      type: "full",
      callback,
    };
    this.subscribers.add(subscriber);
    return () => {
      this.subscribers.delete(subscriber);
    };
  }

  /**
   * Wait until the signal satisfies a predicate. If the predicate is already satisfied, it will
   * return immediately. Otherwise, it will wait until the signal satisfies the predicate.
   */
  public async until(predicate: (data: TValue) => boolean): Promise<TValue> {
    const current = this.get();
    if (predicate(current)) {
      return current;
    }
    const { promise, resolve } = makePromise<TValue>();
    const unsubscribe = this.subscribe(data => {
      if (predicate(data)) {
        resolve(data);
        unsubscribe();
      }
    });
    return await promise;
  }
}

export interface SignalLike<TValue> extends Subscribable<TValue> {
  get(): TValue;
  subscribe(subscriber: Subscriber<TValue>): () => void;
  subscribeFull(subscriber: SignalFullSubscriber<TValue>): () => void;
  pull(): Promise<StripNotAvailable<TValue>> | StripNotAvailable<TValue>;
}

export type WritableSignal<TValue> = readonly [
  signal: SignalLike<TValue>,
  setter: Setter<StripNotAvailable<TValue>>,
];
