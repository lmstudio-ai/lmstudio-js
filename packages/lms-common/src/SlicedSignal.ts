import { applyPatches, type Patch } from "@lmstudio/immer-with-plugins";
import {
  isAvailable,
  LazySignal,
  type NotAvailable,
  type StripNotAvailable,
} from "./LazySignal.js";
import { makeSetterWithPatches, type Setter } from "./makeSetter.js";
import { type SignalLike } from "./Signal.js";

type PathSegment =
  | {
      type: "key";
      key: string | number;
    }
  | {
      type: "mapKey";
      key: any;
    }
  | {
      type: "keyWithDefault";
      key: string | number;
      default: any;
    }
  | {
      type: "mapKeyWithDefault";
      key: any;
      default: any;
    };

/**
 * A sliced signal is a writable signal that represents a portion of another writable signal.
 */
export function makeSlicedSignalFrom<TSource>(
  writableSignal: readonly [signal: SignalLike<TSource>, setter: Setter<TSource>],
) {
  return new SlicedSignalBuilderImpl<
    TSource,
    StripNotAvailable<TSource>,
    TSource extends NotAvailable ? true : false
  >(writableSignal[0], writableSignal[1]);
}

/**
 * Given a value and a path, return the value at the path.
 *
 * At any point, if a default value is used, the defaultUsed callback is called with the index of
 * the path segment where the default value was used.
 */
function drill(value: any, path: Array<PathSegment>, defaultUsed?: (index: number) => void) {
  let current = value;
  let index = 0;
  for (const key of path) {
    if (key.key === "__proto__") {
      throw new Error("Cannot access __proto__");
    }
    switch (key.type) {
      case "key":
        current = current[key.key];
        break;
      case "keyWithDefault": {
        if (current[key.key] === undefined) {
          current = key.default;
          defaultUsed?.(index);
        } else {
          current = current[key.key];
        }
        break;
      }
      case "mapKey":
        current = current.get(key.key);
        break;
      case "mapKeyWithDefault":
        if (!current.has(key.key)) {
          current = key.default;
          defaultUsed?.(index);
        } else {
          current = current.get(key.key);
        }
        break;
      default: {
        const _exhaustiveCheck: never = key;
        return _exhaustiveCheck;
      }
    }
    index++;
  }
  return current;
}

export type PathRelationship = "ancestor" | "children" | "neither";

/**
 * Determine the relationship of a to b.
 *
 * - Returns "ancestor" if a is an ancestor of b.
 * - Returns "children" if
 *   - a is a child of b, or
 *   - a is the same as b.
 * - Returns "neither" if a and b are not direct ancestors or children of each other.
 */
export function pathRelationship(a: Array<string | number>, b: Array<string | number>) {
  if (a.length < b.length) {
    if (pathStartsWith(b, a)) {
      return "ancestor";
    } else {
      return "neither";
    }
  } else {
    if (pathStartsWith(a, b)) {
      return "children";
    } else {
      return "neither";
    }
  }
}

function pathStartsWith(path: Array<string | number>, prefix: Array<string | number>) {
  if (path.length < prefix.length) {
    return false;
  }
  for (let i = 0; i < prefix.length; i++) {
    if (path[i] !== prefix[i]) {
      return false;
    }
  }
  return true;
}

class SlicedSignalBuilderImpl<TSource, TCurrent, TCanBeNotAvailable extends boolean> {
  private readonly accessPath: Array<PathSegment> = [];
  private readonly path: Array<any> = [];
  private readonly tagKey = String(Math.random());
  public constructor(
    private readonly sourceSignal: SignalLike<TSource>,
    private readonly sourceSetter: Setter<TSource>,
  ) {}
  public access<TKey extends keyof TCurrent & (string | number)>(
    key: TKey,
  ): SlicedSignalBuilderImpl<TSource, TCurrent[TKey], TCanBeNotAvailable> {
    this.accessPath.push({ type: "key", key });
    this.path.push(key);
    return this as any;
  }
  public accessWithDefault<TKey extends keyof TCurrent & (string | number)>(
    key: TKey,
    defaultValue: TCurrent[TKey],
  ): SlicedSignalBuilderImpl<TSource, TCurrent[TKey], TCanBeNotAvailable> {
    this.accessPath.push({ type: "keyWithDefault", key, default: defaultValue });
    this.path.push(key);
    return this as any;
  }
  public mapAccess<TKey extends TCurrent extends Map<infer RKey, any> ? RKey : never>(
    key: TKey,
  ): SlicedSignalBuilderImpl<
    TSource,
    TCurrent extends Map<any, infer RValue> ? RValue : never,
    TCanBeNotAvailable
  > {
    this.accessPath.push({ type: "mapKey", key });
    this.path.push(key);
    return this as any;
  }
  public mapAccessWithDefault<
    TKey extends TCurrent extends Map<infer RKey, infer _RValue> ? RKey : never,
  >(
    key: TKey,
    defaultValue: TCurrent extends Map<TKey, infer RValue> ? RValue : never,
  ): SlicedSignalBuilderImpl<
    TSource,
    TCurrent extends Map<TKey, infer RValue> ? RValue : never,
    TCanBeNotAvailable
  > {
    this.accessPath.push({ type: "mapKeyWithDefault", key, default: defaultValue });
    this.path.push(key);
    return this as any;
  }
  public done(): readonly [
    signal: LazySignal<TCurrent | (TCanBeNotAvailable extends true ? NotAvailable : never)>,
    setter: Setter<TCurrent>,
  ] {
    const sourceInitialValue = this.sourceSignal.get();
    const initialValue = isAvailable(sourceInitialValue)
      ? drill(sourceInitialValue, this.accessPath)
      : LazySignal.NOT_AVAILABLE;
    const signal = LazySignal.create<
      TCurrent | (TCanBeNotAvailable extends true ? NotAvailable : never)
    >(initialValue, setDownstream => {
      const unsubscribe = this.sourceSignal.subscribeFull((value, patches, tags) => {
        const newPatches: Array<Patch> = [];
        // Transform patches
        for (const patch of patches) {
          const relationship = pathRelationship(patch.path, this.path);
          // If one of the ancestors have been replaced, we need to replace the whole value
          if (relationship === "ancestor") {
            if (patch.op !== "replace") {
              throw new Error("Only replace patches are supported for ancestor relationships");
            }
            newPatches.length = 0;
            newPatches.push({
              op: "replace",
              path: [],
              value: drill(
                patch.value!,
                // Get the part that is relevant to this slice
                this.accessPath.slice(patch.path.length),
              ),
            });
          } else if (relationship === "children") {
            // If one of the children have been replaced, we need to replace the subset that
            // corresponds to the change
            newPatches.push({
              ...patch,
              path: patch.path.slice(this.accessPath.length),
            });
          }
          // Otherwise, we don't need to do anything
        }
        const newValue = drill(value, this.accessPath);
        const newTags = tags
          .filter(tag => tag.startsWith(this.tagKey))
          .map(tag => tag.slice(this.tagKey.length));
        if (newPatches.length > 0 || newTags.length > 0) {
          setDownstream.withValueAndPatches(newValue, newPatches, newTags);
        }
      });
      const value = this.sourceSignal.pull();
      if (value instanceof Promise) {
        value.then(value => {
          setDownstream(drill(value, this.accessPath));
        });
      } else {
        setDownstream(drill(value, this.accessPath));
      }

      return unsubscribe;
    });
    const setter = makeSetterWithPatches<TCurrent>((updater, tags) => {
      const newTags = tags?.map(tag => this.tagKey + tag);
      this.sourceSetter.withPatchUpdater(oldValue => {
        const newPatches: Array<Patch> = [];
        const slicedOldValue = drill(oldValue, this.accessPath, index => {
          // A default value has been used. We need to inject a patch that adds the default value
          // to the parent object.
          const defaultPathSegment = this.accessPath[index];
          if (
            defaultPathSegment.type !== "keyWithDefault" &&
            defaultPathSegment.type !== "mapKeyWithDefault"
          ) {
            throw new Error("Expected keyWithDefault or mapKeyWithDefault");
          }
          const defaultValue = defaultPathSegment.default;
          newPatches.push({
            op: "replace",
            path: this.path.slice(0, index + 1),
            value: defaultValue,
          });
        });
        const [_newSlicedValue, patches] = updater(slicedOldValue);
        newPatches.push(
          ...patches.map(patch => ({
            ...patch,
            path: [...this.path, ...patch.path],
          })),
        );
        const newValue = applyPatches(oldValue as any, newPatches);
        return [newValue, newPatches] as const;
      }, newTags);
    });
    return [signal, setter];
  }
}

export type SlicedSignalBuilder<
  TSource,
  TCurrent,
  TCanBeNotAvailable extends boolean,
> = SlicedSignalBuilderImpl<TSource, TCurrent, TCanBeNotAvailable>;
