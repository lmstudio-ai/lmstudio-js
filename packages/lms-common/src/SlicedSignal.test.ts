import "."; // Import order

import { LazySignal, makeSetterWithPatches, NotAvailable, Setter } from ".";
import { Signal } from "./Signal.js";
import { makeSlicedSignalFrom, pathRelationship } from "./SlicedSignal.js";

describe("pathRelationship", () => {
  it("should return ancestor", () => {
    const relationship = pathRelationship(["first", "second"], ["first", "second", "third"]);
    expect(relationship).toBe("ancestor");
  });
  it("should return children", () => {
    const relationship = pathRelationship(["first", "second", "third"], ["first", "second"]);
    expect(relationship).toBe("children");
    const relationship2 = pathRelationship(["first", "second"], ["first", "second"]);
    expect(relationship2).toBe("children");
  });
  it("should return neither", () => {
    const relationship = pathRelationship(["first", "second"], ["first", "third"]);
    expect(relationship).toBe("neither");
    const relationship2 = pathRelationship(["first", "weird"], ["first", "second", "third"]);
    expect(relationship2).toBe("neither");
  });
});

describe("SlicedSignal", () => {
  it("should be able to read with regular signal", () => {
    const [sourceSignal, setSource] = Signal.create({ a: { b: { c: 1 } } });
    const [slicedSignal, _setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access("b")
      .done();

    expect(slicedSignal.get()).toEqual({ c: 1 });
  });

  it("should be able to subscribe with regular signal", () => {
    const [sourceSignal, setSource] = Signal.create({ a: { b: { c: 1 } }, d: 5 });
    const [slicedSignal, _setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access("b")
      .done();

    const subscriber = jest.fn();
    const fullSubscriber = jest.fn();
    slicedSignal.subscribe(subscriber);
    slicedSignal.subscribeFull(fullSubscriber);

    expect(subscriber).not.toHaveBeenCalled();
    expect(fullSubscriber).not.toHaveBeenCalled();

    setSource.withProducer(draft => {
      draft.a.b.c = 2;
    });

    expect(subscriber).toHaveBeenCalledWith({ c: 2 });
    expect(fullSubscriber).toHaveBeenCalledWith(
      { c: 2 },
      [
        {
          op: "replace",
          path: ["c"],
          value: 2,
        },
      ],
      [],
    );

    setSource.withProducer(draft => {
      draft.a.b = { c: 3 };
    });

    expect(subscriber).toHaveBeenCalledWith({ c: 3 });
    expect(fullSubscriber).toHaveBeenCalledWith(
      { c: 3 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 3 },
        },
      ],
      [],
    );

    setSource.withProducer(draft => {
      draft.a = { b: { c: 4 } };
    });

    expect(subscriber).toHaveBeenCalledWith({ c: 4 });
    expect(fullSubscriber).toHaveBeenCalledWith(
      { c: 4 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 4 },
        },
      ],
      [],
    );

    expect(subscriber).toHaveBeenCalledTimes(3);
    expect(fullSubscriber).toHaveBeenCalledTimes(3);

    setSource.withProducer(draft => {
      draft.d = 6;
    });

    expect(subscriber).toHaveBeenCalledTimes(3);
    expect(fullSubscriber).toHaveBeenCalledTimes(3);
  });

  it("should be able to set regular signal", () => {
    const [sourceSignal, setSource] = Signal.create({ a: { b: { c: 1 } } });
    const [_slicedSignal, setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access("b")
      .done();

    setSliced({ c: 2 });

    expect(sourceSignal.get()).toEqual({ a: { b: { c: 2 } } });

    setSliced.withProducer(draft => {
      draft.c = 3;
    });

    expect(sourceSignal.get()).toEqual({ a: { b: { c: 3 } } });
  });

  it("should be able to trigger subscription with regular signal", () => {
    const [sourceSignal, setSource] = Signal.create({ a: { b: { c: 1 } } });
    const [slicedSignal, setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access("b")
      .done();

    const sourceSubscriber = jest.fn();
    const slicedSubscriber = jest.fn();
    sourceSignal.subscribeFull(sourceSubscriber);
    slicedSignal.subscribeFull(slicedSubscriber);

    setSliced.withProducer(draft => {
      draft.c = 2;
    });

    expect(sourceSubscriber).toHaveBeenCalledWith(
      { a: { b: { c: 2 } } },
      [
        {
          op: "replace",
          path: ["a", "b", "c"],
          value: 2,
        },
      ],
      [],
    );
    expect(slicedSubscriber).toHaveBeenCalledWith(
      { c: 2 },
      [
        {
          op: "replace",
          path: ["c"],
          value: 2,
        },
      ],
      [],
    );

    setSliced({ c: 3 });

    expect(sourceSubscriber).toHaveBeenCalledWith(
      { a: { b: { c: 3 } } },
      [
        {
          op: "replace",
          path: ["a", "b"],
          value: { c: 3 },
        },
      ],
      [],
    );
    expect(slicedSubscriber).toHaveBeenCalledWith(
      { c: 3 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 3 },
        },
      ],
      [],
    );
  });

  it("should properly preserve tags", () => {
    const [sourceSignal, setSource] = Signal.create({ a: { b: { c: 1 } } });
    const [slicedSignal, setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access("b")
      .done();

    const subscriber = jest.fn();
    slicedSignal.subscribeFull(subscriber);

    setSliced({ c: 2 }, ["tag1", "tag2"]);

    expect(subscriber).toHaveBeenCalledWith(
      { c: 2 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 2 },
        },
      ],
      ["tag1", "tag2"],
    );
  });

  it("should isolate tags from different slices", () => {
    const [sourceSignal, setSource] = Signal.create({ a: { b: { c: 1 } } });
    const [sliced1Signal, setSliced1] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access("b")
      .done();
    const [sliced2Signal, _setSliced2] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access("b")
      .done();

    const subscriber1 = jest.fn();
    sliced1Signal.subscribeFull(subscriber1);

    const subscriber2 = jest.fn();
    sliced2Signal.subscribeFull(subscriber2);

    setSliced1({ c: 2 }, ["tag1", "tag2"]);

    expect(subscriber1).toHaveBeenCalledWith(
      { c: 2 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 2 },
        },
      ],
      ["tag1", "tag2"],
    );

    expect(subscriber2).toHaveBeenCalledWith(
      { c: 2 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 2 },
        },
      ],
      [],
    );
  });

  it("should work with lazy signals", async () => {
    let setDownstream: Setter<{ a: { b: { c: number } } }>;
    const lazy = LazySignal.createWithoutInitialValue<{
      a: { b: { c: number } };
    }>(actualSetDownstream => {
      setDownstream = actualSetDownstream;
      return () => {};
    });

    const setterUpdateFn = jest.fn();

    const setter = makeSetterWithPatches<{ a: { b: { c: number } } } | NotAvailable>(
      setterUpdateFn,
    );

    const [slicedSignal, setSliced] = makeSlicedSignalFrom([lazy, setter])
      .access("a")
      .access("b")
      .done();

    const slicedSubscriber = jest.fn();
    slicedSignal.subscribeFull(slicedSubscriber);

    expect(slicedSubscriber).not.toHaveBeenCalled();
    expect(slicedSignal.get()).toBe(LazySignal.NOT_AVAILABLE);

    const lazySubscriber = jest.fn();
    lazy.subscribeFull(lazySubscriber);

    setDownstream!({ a: { b: { c: 1 } } });

    expect(slicedSubscriber).toHaveBeenCalledWith(
      { c: 1 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 1 },
        },
      ],
      [],
    );

    expect(setterUpdateFn).not.toHaveBeenCalled();

    setSliced.withProducer(draft => {
      (draft as any).c = 2;
    });

    expect(setterUpdateFn).toHaveBeenCalled();
    const updater = setterUpdateFn.mock.calls[0][0];
    expect(updater).toBeInstanceOf(Function);

    setDownstream!.withPatchUpdater(updater);

    expect(lazySubscriber).toHaveBeenCalledWith(
      { a: { b: { c: 2 } } },
      [
        {
          op: "replace",
          path: ["a", "b", "c"],
          value: 2,
        },
      ],
      [],
    );
  });

  it("should be able to read with regular signal with arrays", () => {
    const [sourceSignal, setSource] = Signal.create({ a: [{ b: { c: 1 } }, { d: { e: 2 } }] });
    const [slicedSignal, _setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access(0)
      .access("b")
      .done();

    expect(slicedSignal.get()).toEqual({ c: 1 });
  });

  it("should be able to subscribe with regular signal with arrays", () => {
    const [sourceSignal, setSource] = Signal.create({
      a: [{ b: { c: 1 } }, { d: { e: 2 } }],
      f: 5,
    });
    const [slicedSignal, _setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access(0)
      .access("b")
      .done();

    const subscriber = jest.fn();
    const fullSubscriber = jest.fn();
    slicedSignal.subscribe(subscriber);
    slicedSignal.subscribeFull(fullSubscriber);

    expect(subscriber).not.toHaveBeenCalled();
    expect(fullSubscriber).not.toHaveBeenCalled();

    setSource.withProducer(draft => {
      draft.a[0].b!.c = 2;
    });

    expect(subscriber).toHaveBeenCalledWith({ c: 2 });
    expect(fullSubscriber).toHaveBeenCalledWith(
      { c: 2 },
      [
        {
          op: "replace",
          path: ["c"],
          value: 2,
        },
      ],
      [],
    );

    setSource.withProducer(draft => {
      draft.a[0].b = { c: 3 };
    });

    expect(subscriber).toHaveBeenCalledWith({ c: 3 });
    expect(fullSubscriber).toHaveBeenCalledWith(
      { c: 3 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 3 },
        },
      ],
      [],
    );

    setSource.withProducer(draft => {
      draft.a = [{ b: { c: 4 } }];
    });

    expect(subscriber).toHaveBeenCalledWith({ c: 4 });
    expect(fullSubscriber).toHaveBeenCalledWith(
      { c: 4 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 4 },
        },
      ],
      [],
    );

    expect(subscriber).toHaveBeenCalledTimes(3);
    expect(fullSubscriber).toHaveBeenCalledTimes(3);

    setSource.withProducer(draft => {
      draft.f = 6;
    });

    expect(subscriber).toHaveBeenCalledTimes(3);
    expect(fullSubscriber).toHaveBeenCalledTimes(3);
  });

  it("should be able to set regular signal with arrays", () => {
    const [sourceSignal, setSource] = Signal.create({ a: [{ b: { c: 1 } }, { d: { e: 2 } }] });
    const [_slicedSignal, setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access(0)
      .access("b")
      .done();

    setSliced({ c: 2 });

    expect(sourceSignal.get()).toEqual({ a: [{ b: { c: 2 } }, { d: { e: 2 } }] });

    setSliced.withProducer(draft => {
      draft!.c = 3;
    });

    expect(sourceSignal.get()).toEqual({ a: [{ b: { c: 3 } }, { d: { e: 2 } }] });
  });

  it("should be able to trigger subscription with regular signal with arrays", () => {
    const [sourceSignal, setSource] = Signal.create({ a: [{ b: { c: 1 } }, { d: { e: 2 } }] });
    const [slicedSignal, setSliced] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .access(0)
      .access("b")
      .done();

    const sourceSubscriber = jest.fn();
    const slicedSubscriber = jest.fn();
    sourceSignal.subscribeFull(sourceSubscriber);
    slicedSignal.subscribeFull(slicedSubscriber);

    setSliced.withProducer(draft => {
      draft!.c = 2;
    });

    expect(sourceSubscriber).toHaveBeenCalledWith(
      { a: [{ b: { c: 2 } }, { d: { e: 2 } }] },
      [
        {
          op: "replace",
          path: ["a", 0, "b", "c"],
          value: 2,
        },
      ],
      [],
    );
    expect(slicedSubscriber).toHaveBeenCalledWith(
      { c: 2 },
      [
        {
          op: "replace",
          path: ["c"],
          value: 2,
        },
      ],
      [],
    );

    setSliced({ c: 3 });

    expect(sourceSubscriber).toHaveBeenCalledWith(
      { a: [{ b: { c: 3 } }, { d: { e: 2 } }] },
      [
        {
          op: "replace",
          path: ["a", 0, "b"],
          value: { c: 3 },
        },
      ],
      [],
    );
    expect(slicedSubscriber).toHaveBeenCalledWith(
      { c: 3 },
      [
        {
          op: "replace",
          path: [],
          value: { c: 3 },
        },
      ],
      [],
    );
  });
  it("should be able to read with regular signal with defaults", () => {
    const [sourceSignal, setSource] = Signal.create<{
      a: Record<string, { c: Record<string, { e: number }> }>;
    }>({ a: {} });
    const [sliceSignal, _setSlice] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .accessWithDefault("b", { c: {} })
      .access("c")
      .accessWithDefault("d", { e: 1 })
      .done();

    expect(sliceSignal.get()).toEqual({ e: 1 });
  });
  it("should be able to subscribe with regular signal with defaults", () => {
    const [sourceSignal, setSource] = Signal.create<{
      a: Record<string, { c: Record<string, { e: number }> }>;
    }>({ a: {} });
    const [sliceSignal, setSlice] = makeSlicedSignalFrom([sourceSignal, setSource])
      .access("a")
      .accessWithDefault("b", { c: {} })
      .access("c")
      .accessWithDefault("d", { e: 1 })
      .done();

    const subscriber = jest.fn();
    const fullSubscriber = jest.fn();
    sliceSignal.subscribe(subscriber);
    sliceSignal.subscribeFull(fullSubscriber);

    expect(subscriber).not.toHaveBeenCalled();
    expect(fullSubscriber).not.toHaveBeenCalled();

    setSlice({ e: 2 });

    expect(subscriber).toHaveBeenCalledWith({ e: 2 });
    expect(fullSubscriber).toHaveBeenCalledWith(
      { e: 2 },
      [
        {
          op: "replace",
          path: [],
          value: { e: 1 },
        },
        // A bit unfortunate due to the way we handle defaults. Trigger defaults twice meaning we
        // will receive 1 additional patch.
        {
          op: "replace",
          path: [],
          value: { e: 1 },
        },
        {
          op: "replace",
          path: [],
          value: { e: 2 },
        },
      ],
      [],
    );
  });
});
