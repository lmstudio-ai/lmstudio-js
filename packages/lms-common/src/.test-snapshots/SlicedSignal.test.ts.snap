// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlicedSignal should be able to subscribe with regular signals containing maps 1`] = `
[
  {
    "c": 2,
  },
  [
    {
      "op": "replace",
      "path": [
        "c",
      ],
      "value": 2,
    },
  ],
  [],
]
`;

exports[`SlicedSignal should be able to subscribe with regular signals containing maps 2`] = `
[
  {
    "c": 3,
  },
  [
    {
      "op": "replace",
      "path": [],
      "value": {
        "c": 3,
      },
    },
  ],
  [],
]
`;

exports[`SlicedSignal should be able to subscribe with regular signals containing maps 3`] = `
[
  {
    "c": 4,
  },
  [
    {
      "op": "replace",
      "path": [],
      "value": {
        "c": 4,
      },
    },
  ],
  [],
]
`;
